### ISSUE TITLE
Scaling problem with Tpetra: node of type Openmp

### ISSUE THREAD

**Issue by kvmkrao** [trilinos/Trilinos#1464]
**CC**: @wfspotz 

I built Trilinos with MPI(intel openmpi) as well as Kokkos Openmp backend on stampede2. So, I got the Tpetra  node (#1463) is of type Openmp. 

I wrote a small code to test the performance of Trilinos. In the code, I defined node_type  and passed that to make Tpetra Map, Tpetra sparse matrix, Tpetra multi vector,Tpetra operator and muelu::tpetra operator 
`typedef KokkosClassic::DefaultNode::DefaultNodeType node_type; `

I fixed the number of equations to 8millinos. The maximum number of non-zero entries in a row is seven. I compiled the code and ran with 4 cases with 8 processors.
 
> case1: mpirun -n 8 ./run.exe 
> case2: export OMP_NUM_THREADS=8; mpirun -n 8 ./run.exe 
> case3: export OMP_NUM_THREADS=16; mpirun -n 8 ./run.exe 
> case4: export OMP_NUM_THREADS=24; mpirun -n 8 ./run.exe 

> GMRES solver time in the 4 cases 
> case1: Belos: PseudoBlockGmresSolMgr total solve time          7.761 (1)             
> case2: Belos: PseudoBlockGmresSolMgr total solve time          13.25 (1)             
> case3: Belos: PseudoBlockGmresSolMgr total solve time          18.04 (1)             
> case4: Belos: PseudoBlockGmresSolMgr total solve time           22.33 (1)             

I expected that the solver time will decrease with an increase in the number of threads. But, it is not. Is there anything wrong with this code/solver? 
[code.txt](https://github.com/trilinos/Trilinos/files/1110508/code.txt)

Thank you.

**Comment by aprokop** [comment]
@trilinos/tpetra 

First, your code has multiple identical chunks of code.
Second, the number of equations is not fixed according to this:
```c++
  const size_t numMyElements = 10000000; //map->getNodeNumElements ();
  const global_size_t numGlobalElements = numMyElements *  numProcs ;
```

**Comment by mhoemmen** [comment]
@kvmkrao wrote:
> So, I got the Tpetra node (#1463) is of type Openmp.

Just curious -- what is the value of `Tpetra::Map<>::device_type::execution_space`?

**Comment by kvmkrao** [comment]
@aprokop 
Thank you. 

> First, your code has multiple identical chunks of code.
I fixed some of the identical/unnecessary statements in the code as per your suggestion. Please see the attachment. 
I think, it is better compared to the last time. I compiled this code run some cases. The solver time with those cases for problem size (ndof) of 10M and 20M are as follows: 

_export OMP_NUM_THREADS=$omp
mpirun -n $proc ./run.exe_ 

 case1: ndof=10M, np=4, omp=1     Belos: PseudoBlockGmresSolMgr total solve time  7.075 
case2:`` ndof=10M, np=4, omp=2     Belos: PseudoBlockGmresSolMgr total solve time  5.992 
case3: ndof=10M, np=4, omp=4     Belos: PseudoBlockGmresSolMgr total solve time  5.434 
case4: ndof=10M, np=4, omp=8     Belos: PseudoBlockGmresSolMgr total solve time  5.167 
case5: ndof=10M, np=4, omp=16    Belos: PseudoBlockGmresSolMgr total solve time  5.033 
case6: ndof=10M, np=4, omp=32    Belos: PseudoBlockGmresSolMgr total solve time  6.825 

case7:  ndof=10M, np=8, omp=1    Belos: PseudoBlockGmresSolMgr total solve time  3.587
case8:  ndof=10M, np=8, omp=2    Belos: PseudoBlockGmresSolMgr total solve time  3.052
case9:  ndof=10M, np=8, omp=4    Belos: PseudoBlockGmresSolMgr total solve time  2.783
case10: ndof=10M, np=8, omp=8    Belos: PseudoBlockGmresSolMgr total solve time  4.163
case11: ndof=10M, np=8, omp=16   Belos: PseudoBlockGmresSolMgr total solve time  5.38
case12: ndof=10M, np=8, omp=32   Belos: PseudoBlockGmresSolMgr total solve time  11.04

case13: ndof=10M, np=16, omp=1   Belos: PseudoBlockGmresSolMgr total solve time  1.865
case14: ndof=10M, np=16, omp=2   Belos: PseudoBlockGmresSolMgr total solve time  1.596
case15: ndof=10M, np=16, omp=4   Belos: PseudoBlockGmresSolMgr total solve time  2.303
case16: ndof=10M, np=16, omp=8   Belos: PseudoBlockGmresSolMgr total solve time  5.87
case17: ndof=10M, np=16, omp=16  Belos: PseudoBlockGmresSolMgr total solve time  11.41
case18: ndof=10M, np=16, omp=32  Belos: PseudoBlockGmresSolMgr total solve time  14.68

case19: ndof=20M, np=8, omp=1    Belos: PseudoBlockGmresSolMgr total solve time  7.074 
case20: ndof=20M, np=8, omp=2    Belos: PseudoBlockGmresSolMgr total solve time  6.019 
case21: ndof=20M, np=8, omp=4    Belos: PseudoBlockGmresSolMgr total solve time  5.456 
case22: ndof=20M, np=8, omp=8    Belos: PseudoBlockGmresSolMgr total solve time  7.485 
case23: ndof=20M, np=8, omp=16   Belos: PseudoBlockGmresSolMgr total solve time  8.398 
case24: ndof=20M, np=8, omp=32   Belos: PseudoBlockGmresSolMgr total solve time  14.76 

case25: ndof=20M, np=16, omp =1  Belos: PseudoBlockGmresSolMgr total solve time  3.62            
case26: ndof=20M, np=16, omp =2  Belos: PseudoBlockGmresSolMgr total solve time  3.085           
case27: ndof=20M, np=16, omp =4  Belos: PseudoBlockGmresSolMgr total solve time  4.448           
case28: ndof=20M, np=16, omp =8  Belos: PseudoBlockGmresSolMgr total solve time  8.078           
case29: ndof=20M, np=16, omp =16 Belos: PseudoBlockGmresSolMgr total solve time  16.8            
case30: ndof=20M, np=16, omp =32 Belos: PseudoBlockGmresSolMgr total solve time  21.26   

The solver time is decreasing with an increase in the number of threads. In most of the cases, It is best 
at omp= 4. Can we expect this  type of trend ? can I improve the solver time further?   

> Second, the number of equations is not fixed according to this:
I fixed those lines of code. The number of equations (numGlobalElements) are defined in the code and they are distributed among the processors based on the number of processors. Please see the modified code.
[code_mod.txt](https://github.com/trilinos/Trilinos/files/1112441/code_mod.txt)

Thank you.

**Comment by kvmkrao** [comment]
@mhoemmen 
Thank you

> what is the value of Tpetra::Map<>::device_type::execution_space?

My do-config file contains 
cmake .....
 -D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
  -D Tpetra_INST_OPENMP:BOOL=ON \
 -D Trilinos_ENABLE_Kokkos:BOOL=ON \
 -D Kokkos_ENABLE_OpenMP:BOOL=ON \
     -D Tpetra_ENABLE_Experimental:BOOL=ON \
      -D MueLu_ENABLE_Experimental:BOOL=ON \
      -D Xpetra_ENABLE_Experimental:BOOL=ON \
      -D MueLu_ENABLE_Experimental:BOOL=ON \
      -D Ifpack2_ENABLE_Experimental:BOOL=ON \
      -D TpetraKernels_ENABLE_Experimental:BOOL=ON \
      -D Tpetra_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON \
      -D Xpetra_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON \
      -D Ifpack2_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON \
      -D MueLu_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON \
.........

Therefore, I think It is Kokkos::OpenMP.

However, I am trying  to change the execution space to Kokkos::CUDA.  

If I change the execution space to Kokkos::CUDA, then I will get Tpetra:node of type CUDA.  I checked the documentation for the arguments those should specified during the compilation and/or run time. I was not successful in that. Could you please point me to a proper reference for Tpetra:node (CUDA), if it exists.  (In Kokkos library, we specify the KOKKOS_DEVICES as well KOKKOS_ARCH along with make command.)

Thank you.

**Comment by mhoemmen** [comment]
@kvmkrao wrote:
> `Tpetra_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON`

I'm curious -- where did you get this?  I grepped through all of Trilinos' source code for `EXECUTION_SPACE` and didn't find anything.

I think best practice is to set `Trilinos_ENABLE_OpenMP:BOOL=ON` and rely on the defaults.  OpenMP should be the default for everything downstream.

There is no `Tpetra_ENABLE_Experimental` option.  Did someone tell you to turn all those options on?

My view is that "experimental" options are only for developers.  If package owners find themselves normally telling ordinary users (not a special subset of power users who have a unique relationship with the developers) to enable them, then maybe package owners need to promote those features out of experimental.  Just my opinion, though.

**Comment by mhoemmen** [comment]
@kvmkrao Please execute this code and tell me what it prints:
```
#include "Teuchos_TypeNameTraits.hpp"

std::cout << "Execution space name: ";
typedef Tpetra::Map<>::device_type::execution_space default_execution_space;
std::cout << Teuchos::TypeNameTraits<default_execution_space>::name () << std::endl;
```
You should see "OpenMP" somewhere in that string.  If you see something else, please let me know.  Thanks!

**Comment by kvmkrao** [comment]
> _Tpetra_ENABLE_EXECUTION_SPACE_OPENMP:BOOL=ON_
I'm curious -- where did you get this? I grepped through all of Trilinos' source code for _EXECUTION_SPACE_ and didn't find anything.

I found that in _packages/tpetra/kernels/CMakeLists.txt_. So, I kept those statements in do-config file. I might be interpreted wrongly. 

I am sorry. I kept that statement in the config file. But, forgot to mention that in the issue. Sure, I will do that.  Thanks.

Nobody recommended me that option. I also found the option _(XXXXX_ENABLE_Experimental)_ in _packages/tpetra/kernels/CMakeLists.txt_. Anyway, I read in the documentation that some of the feature are experimental. And thought that this option will enable those experimental features. I understood that this option is meant for developers. 

> You should see "OpenMP" somewhere in that string. If you see something else, please let me know. Thanks!

Thank you. I included those statements and ran the code. I confirm that the execution space is  Kokkos::OpenMP

> output: Execution space name: Kokkos::OpenMP

Thank you.

**Comment by mhoemmen** [comment]
> I found that in packages/tpetra/kernels/CMakeLists.txt. So, I kept those statements in do-config file. I might be interpreted wrongly.

Is this is a release of Trilinos?  tpetra/kernels got replaced with kokkos-kernels several months ago. 

All you need to do is set `Trilinos_ENABLE_OpenMP:BOOL=ON`.  That's it.  Don't set anything else.  Then both Kokkos and Tpetra will use OpenMP by default.

> Anyway, I read in the documentation that some of the feature are experimental. And thought that this option will enable those experimental features. I understood that this option is meant for developers.

Then don't enable them unless some developer tells you to enable them :-) .

> Thank you. I included those statements and ran the code. I confirm that the execution space is Kokkos::OpenMP

Ah, that's good, thanks.

**Comment by jjellio** [comment]
You will see significant performance variability if you are not binding MPI processes to cores.
E.g., for openMPI
mpirun --map-by ppr:2:NUMA:pe=<num> -x OMP_NUM_THREADS=<num> -x OMP_PLACES=cores -x OMP_DISPLAY_ENV=verbose

You should also look at the output adding  --display-map or --report-bindings, OpenMP performance is absolutely awful if don't have proper binding.

Getting binding correct with other job launchers is more complicated.  If OpenMPI's display-map doesn't look right then OpenMPI was not built correctly. (It's hwloc settings are non existent or not correct). I recommend using the --map-by clause and not specifying a specific number of processors, as this will ensure that spawn exactly as many processes as your hardware supports. Mapping by NUMA will also ensure your MPI process does not span multiple NUMA domains.  You will see terrible performance if your threads fight over NUMA.

For example, mpirun --display-map --map-by ppr:2NUMA:pe=7 hostname,  on my desktop:

```
[host] MCW rank 2 bound to socket 1[core 14[hwt 0-1]], socket 1[core 15[hwt 0-1]], socket 1[core 16[hwt 0-1]], socket 1[core 17[hwt 0-1]], socket 1[core 18[hwt 0-1]], socket 1[core 19[hwt 0-1]], socket 1[core 20[hwt 0-1]]: [../../../../../../../../../../../../../..][BB/BB/BB/BB/BB/BB/BB/../../../../../../..]
[host] MCW rank 3 bound to socket 1[core 21[hwt 0-1]], socket 1[core 22[hwt 0-1]], socket 1[core 23[hwt 0-1]], socket 1[core 24[hwt 0-1]], socket 1[core 25[hwt 0-1]], socket 1[core 26[hwt 0-1]], socket 1[core 27[hwt 0-1]]: [../../../../../../../../../../../../../..][../../../../../../../BB/BB/BB/BB/BB/BB/BB]
[host] MCW rank 0 bound to socket 0[core 0[hwt 0-1]], socket 0[core 1[hwt 0-1]], socket 0[core 2[hwt 0-1]], socket 0[core 3[hwt 0-1]], socket 0[core 4[hwt 0-1]], socket 0[core 5[hwt 0-1]], socket 0[core 6[hwt 0-1]]: [BB/BB/BB/BB/BB/BB/BB/../../../../../../..][../../../../../../../../../../../../../..]
[host] MCW rank 1 bound to socket 0[core 7[hwt 0-1]], socket 0[core 8[hwt 0-1]], socket 0[core 9[hwt 0-1]], socket 0[core 10[hwt 0-1]], socket 0[core 11[hwt 0-1]], socket 0[core 12[hwt 0-1]], socket 0[core 13[hwt 0-1]]: [../../../../../../../BB/BB/BB/BB/BB/BB/BB][../../../../../../../../../../../../../..]
```

NUMA domains are enclosed in square braces and bound cores are 'B'. The point being that no process shares a core.

**Comment by kvmkrao** [comment]
@mhoemmen 
> Is this is a release of Trilinos? tpetra/kernels got replaced with kokkos-kernels several months ago.

It is 12.11 version. I think, this version includes the modified kernels. 

> *** Base Git Repo: Trilinos 
> 545269e [Mon Jun 19 16:49:55 2017 -0600] <kyukim@sandia.gov>
> Intrepid2 - check point inclusion

Sure, I don't set/enable them unless I get a word from the developers(you). 

Anyway, I am building Trilinos with the following options to get Tpetra:node type CUDA 

>       -D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
>       -D Trilinos_ENABLE_Tpetra:BOOL=ON \
>       -D Tpetra_ENABLE_Kokkos_Refactor:BOOL=ON \
>       -D Kokkos_ENABLE_Cuda:BOOL=ON \
>       -D Kokkos_ENABLE_Cuda_UVM:BOOL=ON \
>       -D Kokkos_ENABLE_CXX11_DISPATCH_LAMBDA:BOOL=ON \
>       -D TPL_ENABLE_Thrust:BOOL=ON \
>       -D TPL_ENABLE_CUDA:BOOL=ON \
>       -D TPL_ENABLE_CUSPARSE:BOOL=ON \
>       -D Trilinos_ENABLE_Kokkos:BOOL=ON \
>       -D Trilinos_ENABLE_KokkosAlgorithms:BOOL=ON \
>       -D Trilinos_ENABLE_KokkosContainers:BOOL=ON \
>       -D Trilinos_ENABLE_KokkosCore:BOOL=ON \

I tried to set Trilinos_ENABLE_Cuda:BOOL=ON (like Trilinos_ENABLE_OpenMP:BOOL=ON). It did not work. Could you please let me know any option that I am missing in this build to obtain Tpetra:Node CUDA. I could not get the solver time with this build as I get an error related to jobscript (sbatch). 

Thank you.

**Comment by kvmkrao** [comment]
Thank you @jjellio 

I tried to bind the MPI process to cores via the --map-by 
> 
> export OMP_NUM_THREADS=threads
> export OMP_PLACES=cores
> export OMP_DISPLAY_ENV=verbose
> export I_MPI_MIC=1
> 
> mpiexec.hydra --map-by ppr:64:numa:pe=threads -np nproc run.exe

The standard GMRES solver time for various nproc and threads are as follows: 

> problem size(ndof): 20million
> solver: GMRES in Belos, 
> Preconditioner: smoothed aggregation in MueLu          
> 
> threads  nproc=4 nproc=8   nproc=16 nproc=32  nproc=64         
> 1           14.15      7.513        3.626        1.9           1.077         
> 2           11.91      6.013       3.096         1.637       1.56
> 4           10.81      5.498       4.501         2.402       2.19
> 8           10.24      7.588       7.977         3.789       7.342
> 16          10.24    8.415       16.74          10.24       16.21
> 32         10.28    15.54        22.86          18.28       30.39

I think, i am doing the binding properly for this case.  

> Getting binding correct with other job launchers is more complicated. If OpenMPI's display-map doesn't look right then OpenMPI was not built correctly. (It's hwloc settings are non existent or not correct). I recommend using the --map-by clause and not specifying a specific number of processors, as this will ensure that spawn exactly as many processes as your hardware supports. Mapping by NUMA will also ensure your MPI process does not span multiple NUMA domains. You will see terrible performance if your threads fight over NUMA.

Thank you very much.

**Comment by mhoemmen** [comment]
> `-D Tpetra_ENABLE_Kokkos_Refactor:BOOL=ON`

There is no need to set this.  It is ON by default.  

> `-D TPL_ENABLE_Thrust:BOOL=ON`

Tpetra does not use or need Thrust.

> `Trilinos_ENABLE_Kokkos:BOOL=ON`

If you enable Tpetra, that will enable Kokkos, KokkosCore, KokkosAlgorithms, and KokkosContainers.  You do not need to enable Kokkos* explicitly unless you plan on running their unit tests or examples.

**Comment by mhoemmen** [comment]
Here is how you do a CUDA build:

Trilinos currently tests CUDA builds only with the OpenMPI version of MPI.  We recommend OpenMPI 1.8.x at a minimum, preferably newer. 

First, set the following environment variables, in which `TRILINOS_PATH` points to your Trilinos source directory:
```
  export OMPI_CXX=${TRILINOS_PATH}/packages/kokkos/bin/nvcc_wrapper
  export CUDA_LAUNCH_BLOCKING=1
  export CUDA_MANAGED_FORCE_DEVICE_ALLOC=1
```
Then set the following CMake variables, in addition to the CMake variables that you would normally set:
```
-D TPL_ENABLE_MPI:BOOL=ON
-D TPL_ENABLE_CUDA:BOOL=ON
-D Kokkos_ENABLE_Cuda_UVM:BOOL=ON
-D Tpetra_INST_CUDA:BOOL=ON
```
If the version of `mpicxx` that lives in your path is correct, then this should work.

**Comment by kvmkrao** [comment]
Thank you @mhoemmen 

On Pittsburgh supercomputer, I loaded openmpi-1.10.2, cuda/8.0RC modules. The redundant options in the configuration file is removed as per your suggestions. I included "Tpetra_INST_CUDA:BOOL=ON" in my configuration as it was missing. 

>   export OMPI_CXX=${TRILINOS_PATH}/packages/kokkos/bin/nvcc_wrapper
  export CUDA_LAUNCH_BLOCKING=1
  export CUDA_MANAGED_FORCE_DEVICE_ALLOC=1

The supercomputer gives us access to Tesla K80 and P100 GPUs. Therefore, I changed the default arch to _sm_60_  (for P100 GPUs use) and set _OMPI_CXX_ as well as CUDA options. 
Also set the NVCC_WRAPPER_DEFAULT_COMPILER to g++

> If the version of mpicxx that lives in your path is correct, then this should work.
The path in which mpicxx lives is set in my bashrc file. Thanks. 

I am able to compiled and run the first two examples (01_init, 02_problem) at packages/tpetra/core/example/Lesson07-Kokkos-Fill/. I could not compile the remaining examples without errors.

The error while compiling _03_fill_ example is 
> /tmp/tmpxft_0000274f_00000000-4_03_fill.cudafe1.stub.c:41:381:   required from here
>/pylon2/ac4s8rp/kotteda/cuda/Software/Trilinos/install1/include/impl/Kokkos_FunctorAdapter.hpp:395:76: error: no matching function for call to ‘Kokkos::Impl::FunctorValueTraits<__nv_hdl_wrapper_t<__nv_dl_tag<int (*)(int, char**), main, 3u>, void(int, long unsigned int&, bool), Kokkos::View<long unsigned int*>, int, Kokkos::View<long unsigned int*> >, void, false>::deduce_reduce_type(Kokkos::Impl::FunctorValueTraits<__nv_hdl_wrapper_t<__nv_dl_tag<int (*)(int, char**), main, 3u>, void(int, long unsigned int&, bool), Kokkos::View<long unsigned int*>, int, Kokkos::View<long unsigned int*> >, void, false>::tag_type, <unresolved overloaded function type>)’
>    typedef decltype( deduce_reduce_type( tag_type() , & FunctorType::operator() ) ) ValueType ; 

I request your suggestion to  eliminate  this compilation error. 

Could you please point me to an example/test which is suitable for this installation as well as gives me an idea to fill the sparse matrix and solve the linear system. 

Thank you very much.

**Comment by mhoemmen** [comment]
> Therefore, I changed the default arch to sm_60 (for P100 GPUs use)

How did you do that?

**Comment by mhoemmen** [comment]
Also, are you able to build Kokkos independently of Trilinos, using its Makefile-based build system?  Please try that first.  You can get Kokkos at github.com/kokkos/kokkos and build its tests and examples.  If you can get that to work, then we might have some chance of Trilinos working.

**Comment by kvmkrao** [comment]
> How did you do that?
I changed that in packages/kokkos/config/nvcc_wrapper 

> $head packages/kokkos/config/nvcc_wrapper 
> # Default settings: change those according to your machine.  For
> # example, you may have have two different wrappers with either icpc
> # or g++ as their back-end compiler.  The defaults can be overwritten
> # by using the usual arguments (e.g., -arch=sm_30 -ccbin icpc).
> 
> default_arch="sm_60"

Also, are you able to build Kokkos independently of Trilinos, using its Makefile-based build system? Please try that first. You can get Kokkos at github.com/kokkos/kokkos and build its tests and examples. If you can get that to work, then we might have some chance of Trilinos working.

Thank you @mhoemmen 
I cloned the kokkos repository from github and installed it on the computer using the following commands
$./generate_makefile.bash --prefix=$path/kokkos/install
$make -j KOKKOS_DEVICES=Cuda
$make install 

Then, I compiled the examples at kokkos/example/tutorial using the makefile  
_make -j KOKKOS_DEVICES=Cuda_
I considered example2 (simple_reduce) and compiled it using the makefile. The output  is 

> Kokkos::Cuda::initialize WARNING: running kernels compiled for compute capability 3.5 on device with compute capability 3.7 , this will likely reduce potential performance.
> Kokkos::Cuda::initialize WARNING: running kernels compiled for compute capability 3.5 on device with compute capability 3.7 , this will likely reduce potential performance.
> Sum of squares of integers from 0 to 9, computed in parallel, is 285
> Sum of squares of integers from 0 to 9, computed sequentially, is 285
> Sum of squares of integers from 0 to 9, computed in parallel, is 285
> Sum of squares of integers from 0 to 9, computed sequentially, is 285

Please let me know, if I there are any corrections.        
Thank you.

**Comment by mhoemmen** [comment]
> I changed that in packages/kokkos/config/nvcc_wrapper

Could you try reconfiguring and rebuilding Trilinos without changing anything in `nvcc_wrapper`?

**Comment by jjellio** [comment]
@kvmkrao 

`mpiexec.hydra --map-by ppr:64:numa:pe=threads -np nproc run.exe`

If mpiexec.hydra is OpenMPI, then the above line would be correct if you intend to have 64 MPI processes per NUMA domain, and each process will have 'pe' number of threads. I doubt this is what you want, perhaps use 'sockets' instead of NUMA. 'ppr' is the number of MPI processes that will be spawned inside the following resource (numa in this case). Each process will have 'pe' number of processors assigned to it.  As pe gets larger, the ppr part has to get smaller or you are oversubscribing the node.

I'd also recommend experimenting with GMRES only. MueLu is not fully threaded.

**Comment by kvmkrao** [comment]
Thank you @mhoemmen 
I cloned the Trilinos git repository.  The repo infomation is as follows: 
--------------------------------------------------------------------------------
*** Base Git Repo: Trilinos  | 4c8adc9 [Mon Jun 26 16:12:09 2017 -0600] <lberge@sandia.gov> | MueLu: adding stokhos files
 --------------------------------------------------------------------------------

I used the modified configuration file and built the trilinos.

module load cmake
module load pgi/17.5
module load mpi/pgi_openmpi
module load cuda/8.0RC

The configuration file contains the following options 

export NVCC_WRAPPER_DEFAULT_COMPILER=/usr/lib64/ccache/g++
export OMPI_CXX=/pylon2/ac4s8rp/kotteda/cuda/Software/Trilinos/gitsrc/packages/kokkos/config/nvcc_wrapper
#(Please not that I did not change the default nvcc wrapper for this build)
export CUDA_LAUNCH_BLOCKING=1
export CUDA_MANAGED_FORCE_DEVICE_ALLOC=1

>cmake -D CMAKE_BUILD_TYPE:STRING=DEBUG \
>      -D CMAKE_C_COMPILER="mpicc" \
>      -D CMAKE_CXX_COMPILER="mpicxx" \
>      -D CMAKE_Fortran_COMPILER="mpif90" \
      -D CMAKE_CXX_FLAGS="-std=c++11 --expt-extended-lambda -DKOKKOS_CUDA_USE_LAMBDA" \
>     -D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON \
>     -D Trilinos_ENABLE_DEBUG:BOOL=OFF \
>     -D Trilinos_ENABLE_INSTALL_CMAKE_CONFIG_FILES:BOOL=ON \
>      -D TPL_ENABLE_MPI:BOOL=ON \
>\
>      -D Trilinos_ASSERT_MISSING_PACKAGES:BOOL=OFF \
>      -D Trilinos_ENABLE_ALL_PACKAGES:BOOL=OFF \
>      -D Trilinos_WARNINGS_AS_ERRORS_FLAGS:STRING="" \
>\
>      -D Trilinos_ENABLE_CXX11:BOOL=ON \
>      -D Trilinos_ENABLE_Tpetra:BOOL=ON \
>      -D Trilinos_ENABLE_Belos:BOOL=ON \
>      -D Trilinos_ENABLE_Ifpack2:BOOL=ON \
>      -D Trilinos_ENABLE_MueLu:BOOL=ON \
>      -D Trilinos_ENABLE_TESTS:BOOL=OFF \
>      -D Trilinos_ENABLE_Teuchos:BOOL=ON \
>\
>      -D Trilinos_ENABLE_OpenMP:BOOL=OFF \
>      -D Tpetra_INST_CUDA:BOOL=ON \
>      -D TPL_ENABLE_CUDA:BOOL=ON \
>      -D TPL_ENABLE_CUSPARSE:BOOL=ON \
>      -D Kokkos_ENABLE_Cuda:BOOL=ON \
>      -D Kokkos_ENABLE_Cuda_UVM:BOOL=ON \
>      -D Kokkos_ENABLE_OpenMP:BOOL=OFF \
>      -D Kokkos_ENABLE_Pthread:BOOL=OFF \
>/pylon2/ac4s8rp/kotteda/cuda/Software/Trilinos/gitsrc

 And, the errors while building the Trilinos are 
> Scanning dependencies of target kokkoscore
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_2.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_1.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_control.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_dump.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_global.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_info.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l1.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_aat.c.o
> [  1%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_defaults.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_order.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_post_tree.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_aat.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_postorder.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_preprocess.c.o
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l2.c.o
> pgcc-Error-Unknown switch: -Wno-allpgcc-Error-
> Unknown switch: -Wno-allpgcc-Error-
> pgcc-Error-Unknown switch: -Wno-allpgcc-Error-Unknown switch: -Wno-all
> Unknown switch: -Wno-all
> 
> pgcc-Error-pgcc-Error-Unknown switch: -Wno-allUnknown switch: -Wno-pedanticpgcc-Error-
> 
> Unknown switch: -Wno-pedantic
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-pgcc-Error-Unknown switch: -Wno-pedanticUnknown switch: -Wno-pedantic
> 
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-Unknown switch: -Wno-all
> pgcc-Error-Unknown switch: -Wno-allpgcc-Error-
> Unknown switch: -Wno-all
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-pgcc-Error-pgcc-Error-Unknown switch: -Wno-allpgcc-Error-Unknown switch: -Wno-allpgcc-Error-
> pgcc-Error-Unknown switch: -Wno-all
> Unknown switch: -Wno-pedanticpgcc-Error-Unknown switch: -Wno-pedanticUnknown switch: -Wno-all
> 
> Unknown switch: -Wno-all
> 
> 
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-Unknown switch: -Wno-pedantic
> pgcc-Error-Unknown switch: -Wno-allpgcc-Error-pgcc-Error-
> Unknown switch: -Wno-pedanticUnknown switch: -Wno-pedantic
> 
> pgcc-Error-Unknown switch: -Wno-pedantic
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_2.c.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_dump.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_1.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_control.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_global.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_info.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l1.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_aat.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_defaults.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_order.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_post_tree.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_aat.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_postorder.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l_preprocess.c.o] Error 1
> make[2]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l2.c.o] Error 1
> make[1]: *** [packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> 
> [  2%] Building C object packages/common/auxiliarySoftware/SuiteSparse/src/CMakeFiles/trilinosss.dir/AMD/Source/trilinos_amd_l2.c.o

I could not figure out a way to remove those switches which are causing the errors. 
I appreciate your advice on that?  

Thank you.

**Comment by mhoemmen** [comment]
@kvmkrao It looks like you are using the PGI compiler.  Could you switch to GCC and try again?  Trilinos does not currently have any nightly tests for PGI + CUDA.

**Comment by kvmkrao** [comment]
@mhoemmen 
I am using pgi compiler. It was recommended on the supercomputer user guide <https://www.psc.edu/bridges/user-guide/gpu-use> 
Thank you. 

Anyway, I will use the configuration which is mentioned in the previous comment for rebuilding Trilinos with GCC + CUDA. The following modules will be used  

module load cuda/8.0
module load cmake # cmake version 3.8.2
module load gnu  # gcc version 4.9.2 
module load openmpi_ib/1.8.4 

The errors while building Trilinos are due to compiler flags in Amesos package. And, I get rid of them by disabling Amesos package.  Therefore, my build is successful. 
> packages/amesos/src/CMakeLists.txt:   COMPILE_FLAGS "-Wno-all -Wno-pedantic")

Thanks

**Comment by mhoemmen** [comment]
@kvmkrao Do you need to use the PGI compiler?  If so, we can file a separate issue for that.  Otherwise, I will close this issue.  Thanks!

**Comment by kvmkrao** [comment]
Thank you @jjellio 
I think, it is intel MPI library. 

> mpiicpc for the Intel(R) MPI Library 2017 Update 3 for Linux*
> Copyright(C) 2003-2017, Intel Corporation.  All rights reserved.
> icpc version 17.0.4 (gcc version 5.4.0 compatibility)

> As pe gets larger, the ppr part has to get smaller or you are oversubscribing the node.
You comment made me to realize that I am oversubscribing the node. 
The compute node on the cluster has 4 sockets (68 cores per socket) and 272 cores. The number of cores on the socket is constant. So, I consider appropriate number of MPI processors (ppr) and pe number of processors assigned to it. I will change the mapping from NUMA to socket as per your advice.

Thank you very much.

**Comment by jjellio** [comment]
It sounds like you are running on a KNL node that uses subnuma clustering (SNC=4).  I normally run with OMP_PLACES=cores(num_cores), and OMP_PROC_BIND=spread. The map-by stuff with OpenMPI or any job launcher is more about making sure you do not oversubscribe the node. The environment variables OMP_PLACES and OMP_PROC_BIND then maps the OpenMP threads to the chunk of cores OpenMPI (or srun) allocated to that process. So you need to make sure that OpenMP threads are bound correctly as well. OMP_DISPLAY_ENV=verbose will show you enough information to verify this.

For KNL, you also want to inspect the verbose output to make sure that OMP_WAIT_POLICY is set to passive if you have more than one thread per core (KNL exposes 4 HT per core).  Getting good performance on KNL is not easy. You also want to make sure that the environment variables are being propagated to all MPI processes. For OpenMPI that is done with '-x OMP_NUM_THREADS'

**Comment by mhoemmen** [comment]
btw I am closing this issue but please feel free to continue discussion.  Thanks!
